test/modules/layers/test_attention.py

```diff
@@ -17,164 +19,300 @@
 )
 
 
-class TestAttention(unittest.TestCase):
-    """
-    Test all Attention classes
-    """
+@pytest.fixture(autouse=True)
+def set_seed():
+    set_rng_seed(4)
+
+
+@pytest.fixture
+def hidden_dim():
+    return 3
+
+
+@pytest.fixture
+def n_heads():
+    return 1
+
+
+@pytest.fixture
+def n_dim():
+    return 3
+
+
+@pytest.fixture
+def input_shape(n_dim):
+    return tuple(repeat(2, n_dim))
+
+
+@pytest.fixture
+def q(n_heads, input_shape, hidden_dim):
+    return torch.randn(1, n_heads, *input_shape, hidden_dim // n_heads)
+
+
+@pytest.fixture
+def kv(n_heads, input_shape, hidden_dim):
+    return torch.randn(1, n_heads, *input_shape, hidden_dim // n_heads)
+
+
+@pytest.fixture
+def full_attn(input_shape):
+    # TODO: retire causal once mask generation is moved out of FullAttention
+    #   causal inside FullAttention does not affect caching of k, v
+    return FullAttention(input_shape, causal=False, attn_dropout=0.0)
+
+
+@pytest.fixture
+def axial_attn():
+    return AxialAttention(1)  # only on second axis of input
 
-    def setUp(self):
-        set_rng_seed(4)
-        self.hidden_dim = 3
-        self.n_heads = 1
-        self.input_shape = (2, 2, 2)
-        self.q = torch.randn(
-            1, self.n_heads, *self.input_shape, self.hidden_dim // self.n_heads
-        )
-        self.k = torch.randn(
-            1, self.n_heads, *self.input_shape, self.hidden_dim // self.n_heads
-        )
-        self.v = torch.randn(
-            1, self.n_heads, *self.input_shape, self.hidden_dim // self.n_heads
-        )
-        self.full = FullAttention(self.input_shape, causal=False, attn_dropout=0.0)
-        self.ax = AxialAttention(1)  # only on second axis of input
-        self.mha = MultiHeadAttention(
-            self.input_shape,
-            self.hidden_dim,
-            self.hidden_dim,
-            self.n_heads,
-            1,
-            causal=False,
-            attn_module=self.full,
-        )
-        self.block = AxialAttentionBlock(
-            len(self.input_shape), self.hidden_dim, self.n_heads
-        )
 
-    def test_scaled_dot_product_attention(self):
-        actual = scaled_dot_product_attention(self.q, self.k, self.v)
+class TestMultiheadAttention:
+    @pytest.fixture
+    def multihead_attn(self, input_shape, hidden_dim, n_heads):
+        def create_multihead_attn(causal, attn_module):
+            return MultiHeadAttention(
+                input_shape, hidden_dim, hidden_dim, n_heads, 1, causal, attn_module
+            )
+
+        return create_multihead_attn
+
+    def test_split_multihead(self, input_shape, multihead_attn, full_attn):
+        mha = multihead_attn(False, full_attn)
+        x = torch.randn(1, *input_shape, 6)  # (b, d1, ..., dn, c)
+        mha.n_head = 2
```

Comment:
you could make `n_head` a parameter in `create_multihead_attn` so you can remove fixture and you don't have to modify it here